compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -203,10 +203,12 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         // type/region parameters.
         let self_ty = obligation.self_ty().skip_binder();
         match self_ty.kind() {
-            ty::Generator(..) => {
-                debug!(?self_ty, ?obligation, "assemble_generator_candidates",);
+            ty::Generator(did, ..) => {
+                if !self.tcx().generator_is_async(*did) {
```

Comment:
Can you move this into the match pattern as a `if` guard?

Also a comment could be useful here explaining why this check is particularly important.